rfc/bulksave.md

```diff
@@ -0,0 +1,220 @@
+# Improve the sql storage performance by using bulk save api
+
+| | |
+|-|-|
+| **Status**     | _**Proposed**, Accepted, Implemented, Obsolete_ |
+| **RFC #**      | https://github.com/spinnaker/gate/pull/8368 | https://github.com/spinnaker/orca/pull/3763 | https://github.com/spinnaker/front50/pull/3763
+| **Author(s)**  | Sanjeev Thatiparthi (@sanopsmx) (https://github.com/sanopsmx)
+| **SIG / WG**   | sig-security
+
+## Overview
+
+https://github.com/spinnaker/spinnaker/issues/6147
+
+ This is valid only for sql storage.
+
+ When we have multiple pipelines to save , we need to call front50 save api serially one after the other.
+ This reduces the performance of front50 save api.
+
+ We improve the front50 save performance by using bulk insert query.
+ This will help save many pipelines at once which greatly improves the performance of front50.
+ We need to write a bulk save api which accepts an list of pipelines in a single rest api call
+ and save them in front50 using bulk insert query.
+
+### Goals and Non-Goals
+
+ We improve the front50 save performance by using bulk insert query.
+ This will help save many pipelines at once which greatly improves the performance of front50.
+ We need to write a bulk save api which accepts an list of pipelines in a single rest api call
+ and save them in front50 using bulk insert query.
+
+## Motivation and Rationale
+
+To improve the performance of multiple saving of pipelines at once using bulk insert sql query.
+
+## Timeline
+
+Already Implemented
+
+## Design
+
+1. Deck calls the gate api:
+
+   1.1	url: 		http://<ip address>:8084/pipelines/bulksave
+   method: 	POST
+   payload:	[{"keepWaitingPipelines":false,"limitConcurrent":true,"application":"apptest","spelEvaluator":"v4",
+   "lastModifiedBy":"admin","name":"pipeline_test","stages":[{"refId":"1","requisiteStageRefIds":[],
+   "type":"wait","name":"Wait","waitTime":30}],"index":3,"id":"66a5c77b-3998-495d-ad61-117313c5800c",
+   "triggers":[],"updateTs":"1599652266000"},{"keepWaitingPipelines":false,"limitConcurrent":true,
+   "application":"apptest","spelEvaluator":"v4","lastModifiedBy":"admin","name":"pipeline_test1",
+   "stages":[{"refId":"1","requisiteStageRefIds":[],"type":"wait","name":"Wait","waitTime":30}],
+   "index":4,"id":"66a5c77b-3998-495d-ad61-117313c5800c","triggers":[],"updateTs":"1599652266000"}......]
+
+
+2. Gate api flow:
+
+   2.1	The post request sent by the deck will hit the PipelineController Rest Controller endpoint.
+
+        com.netflix.spinnaker.gate.controllers.PipelineController.bulksavePipeline(Body pipeline..)
```

Comment:
Not sure why we need a dedicated endpoint for this ... can it not just be submitted asynchronously as a task a polled.

We should avoid the `taskService.createAndWaitForCompletion()` wheverer possible.

Reply:
We kept the dedicate endpoint for a rest call. Was able to submit 3000+ pipelines thru json payload and got the response time ~2 mins. Hence, made the call as synchronous. It can be called from deck as well.

Reply:
We are adding only 2 rest endpoints. One in gate and the other in front50. Not adding any rest endpoint in orca.